cmake/modules/AddSwift.cmake

```diff
@@ -355,11 +394,25 @@ function(_add_variant_link_flags)
       endif()
   endif()
 
-  if(NOT "${SWIFT_${LFLAGS_SDK}_ICU_UC}" STREQUAL "")
-    list(APPEND result "-L${SWIFT_${sdk}_ICU_UC}")
-  endif()
-  if(NOT "${SWIFT_${LFLAGS_SDK}_ICU_I18N}" STREQUAL "")
-    list(APPEND result "-L${SWIFT_${sdk}_ICU_I18N}")
+  # The MSVC linker has different flags to other linkers.
+  # TODO: we should consider moving the non-MSVC code to use link_directories
+  # as this reduces code duplication.
+  if ("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
+    if(NOT "${SWIFT_${LFLAGS_SDK}_ICU_UC_LIBRARIES}" STREQUAL "")
+      link_directories("${SWIFT_${sdk}_ICU_UC_LIBRARIES}")
+    endif()
+
+    if(NOT "${SWIFT_${LFLAGS_SDK}_ICU_I18N_LIBRARIES}" STREQUAL "${SWIFT_${LFLAGS_SDK}_ICU_UC_LIBRARIES}")
+      link_directories("${SWIFT_${sdk}_ICU_UC_LIBRARIES}")
+    endif()
+  else()
+    if(NOT "${SWIFT_${LFLAGS_SDK}_ICU_UC}" STREQUAL "")
+      list(APPEND result "-L${SWIFT_${sdk}_ICU_UC}")
+    endif()
+
+    if(NOT "${SWIFT_${LFLAGS_SDK}_ICU_I18N}" STREQUAL "")
+      list(APPEND result "-L${SWIFT_${sdk}_ICU_I18N}")
+    endif()
   endif()
```

Comment:
I would invert the conditions here, pushing the MSVC check inside the two clauses since the difference is the spelling.